packages/replayio/src/utils/createPromiseQueue.ts

```diff
@@ -0,0 +1,91 @@
+import { Deferred, createDeferred } from "./createDeferred";
+
+interface PromiseQueueRoot {
+  add: <T>(action: () => Promise<T>) => Promise<T>;
+}
+
+interface PromiseQueue extends PromiseQueueRoot {
+  fork: () => PromiseQueue;
+  waitUntilIdle: () => Promise<undefined>;
+}
+
+function createGroup(root: PromiseQueueRoot): PromiseQueue {
```

Comment:
This kind of code really benefits from having unit tests. I think we should either add them now or in a fast-follow commit.

Reply:
Yeah, I have that as a TODO here and I want to add them asap. I tested this manually and stuff but I'm still anxious about this change. I will feel way better about landing this with those tests in. So my plan is to add them tomorrow morning and *then* land this and the one stacked on top of it.

Reply:
I dig it.

Reply:
I pushed out tests for this, feel free to take a look at them if you want. I'll land this once CI gets green